import sys
from PyQt6 import QtWidgets, uic, QtCore, QtGui 
from google import genai
from google.genai.errors import APIError
from google.genai import types
import os
from gtts import gTTS
from playsound import playsound
import uuid 
from dotenv import load_dotenv 

# =================================================================
# ğŸŒŸ ë°ì´í„°ë² ì´ìŠ¤ ê´€ë ¨ ëª¨ë“ˆ ì¶”ê°€
# =================================================================
import pymysql
from datetime import datetime

# =================================================================
# 1. ì„¤ì • ë° ì´ˆê¸°í™”
# =================================================================

load_dotenv() 
UI_FILE_NAME = "Mygemini.ui"
GEMINI_API_KEY = os.environ.get("GEMINI_API_KEY") 

# 2. Gemini í´ë¼ì´ì–¸íŠ¸ ì´ˆê¸°í™”
if not GEMINI_API_KEY:
    print("ì¹˜ëª…ì  ì˜¤ë¥˜: GEMINI_API_KEYê°€ í™˜ê²½ ë³€ìˆ˜ë‚˜ .env íŒŒì¼ë¡œ ì„¤ì •ë˜ì§€ ì•Šì•˜ìŠµë‹ˆë‹¤.")
    sys.exit(1)
    
try:
    client = genai.Client(api_key=GEMINI_API_KEY)
    MODEL_NAME = "gemini-2.5-flash"
except ValueError as e:
    print(f"í´ë¼ì´ì–¸íŠ¸ ì´ˆê¸°í™” ì˜¤ë¥˜: API í‚¤ê°€ ì˜ëª»ë˜ì—ˆê±°ë‚˜ ëˆ„ë½ë˜ì—ˆìŠµë‹ˆë‹¤. {e}")
    sys.exit(1)


# =================================================================
# 2. Gemini API í˜¸ì¶œì„ ìœ„í•œ ì›Œì»¤ ìŠ¤ë ˆë“œ (QThread)
# =================================================================

class GeminiWorker(QtCore.QThread):
    response_ready = QtCore.pyqtSignal(str, str) 
    error_occurred = QtCore.pyqtSignal(str, str) 

    def __init__(self, client, model_name, chat_history, user_question):
        super().__init__()
        self.client = client
        self.model_name = model_name
        self.chat_history = chat_history 
        self.user_question = user_question
        
        self.config = types.GenerateContentConfig(
            system_instruction="You are a helpful assistant. Please answer all questions in Korean."
        )

    def run(self):
        try:
            response = self.client.models.generate_content(
                model=self.model_name,
                contents=self.chat_history,
                config=self.config
            )
            
            gemini_response = response.text
            
            self.chat_history.append({"role": "model", "parts": [{"text": gemini_response}]})

            self.response_ready.emit(self.user_question, gemini_response)

        except APIError as e:
            self.error_occurred.emit("API ì˜¤ë¥˜", f"Gemini ì„œë²„ì— ì—°ê²°í•  ìˆ˜ ì—†ìŠµë‹ˆë‹¤. (ì˜¤ë¥˜: {e})")
        except Exception as e:
            self.error_occurred.emit("ì˜ˆê¸°ì¹˜ ì•Šì€ ì˜¤ë¥˜", str(e))

# =================================================================
# ğŸŒŸ 2-2. TTS/Playsound ì²˜ë¦¬ë¥¼ ìœ„í•œ ì›Œì»¤ ìŠ¤ë ˆë“œ (QThread) - ë¸”ë¡œí‚¹ ë¬¸ì œ í•´ê²°
# =================================================================

class SpeechWorker(QtCore.QThread):
    """
    gTTS/playsoundëŠ” ë¸”ë¡œí‚¹ ì‘ì—…ì´ë¯€ë¡œ ë³„ë„ ìŠ¤ë ˆë“œì—ì„œ ì‹¤í–‰í•˜ì—¬ UI ë©ˆì¶¤ì„ ë°©ì§€í•©ë‹ˆë‹¤.
    """
    def __init__(self, text):
        super().__init__()
        self.text = text

    def run(self):
        try:
            print("ìŒì„± ìƒì„± ë° ì¬ìƒ ì‹œì‘ (SpeechWorker ìŠ¤ë ˆë“œ)...")
            filename = f"temp_speech_{uuid.uuid4().hex}.mp3"
            
            tts = gTTS(text=self.text, lang='ko')
            tts.save(filename)
            playsound(filename)
            os.remove(filename)
            print("ìŒì„± ì¬ìƒ ì™„ë£Œ ë° íŒŒì¼ ì‚­ì œ.")
            
        except Exception as e:
            print(f"ìŒì„± ì¶œë ¥ ì˜¤ë¥˜ (gTTS/playsound) (ìŠ¤ë ˆë“œ): {e}")


# =================================================================
# 3. ë©”ì¸ ì• í”Œë¦¬ì¼€ì´ì…˜ í´ë˜ìŠ¤
# =================================================================

class GeminiApp(QtWidgets.QDialog):
    def __init__(self):
        super().__init__()
        
        try:
            uic.loadUi(UI_FILE_NAME, self)
        except FileNotFoundError:
            print(f"ì˜¤ë¥˜: {UI_FILE_NAME} íŒŒì¼ì„ ì°¾ì„ ìˆ˜ ì—†ìŠµë‹ˆë‹¤. ê²½ë¡œë¥¼ í™•ì¸í•´ ì£¼ì„¸ìš”.")
            sys.exit(1)
            
        self.setWindowTitle(f"Gemini Q&A - ëª¨ë¸: {MODEL_NAME}")
        
        self.btnSend = self.findChild(QtWidgets.QPushButton, 'btnSend')
        self.lineEditMyQuestion = self.findChild(QtWidgets.QLineEdit, 'lineEditMyQuestion')
        self.lblAnswer = self.findChild(QtWidgets.QTextEdit, 'lblAnswer') 
        
        if not self.lblAnswer:
            self.lblAnswer = self.findChild(QtWidgets.QPlainTextEdit, 'lblAnswer')
        
        if not self.btnSend or not self.lblAnswer or not self.lineEditMyQuestion:
            print("ì¹˜ëª…ì  ì˜¤ë¥˜: UI íŒŒì¼ì—ì„œ í•„ìˆ˜ ìœ„ì ¯ì„ ì°¾ì„ ìˆ˜ ì—†ìŠµë‹ˆë‹¤.")
            sys.exit(1)

        # ğŸŒŸ UIì— ê²€ìƒ‰ ê¸°ëŠ¥ì„ ì—°ê²°í•˜ê¸° ìœ„í•œ ì„ì‹œ ë²„íŠ¼ ìƒì„± ë° ì—°ê²° (UI íŒŒì¼ì— ë²„íŠ¼ ì¶”ê°€ê°€ í•„ìš”í•©ë‹ˆë‹¤.)
        # self.btnSearch = self.findChild(QtWidgets.QPushButton, 'btnSearch') 
        # if self.btnSearch:
        #     self.btnSearch.clicked.connect(self.prompt_and_search_history)
        
        # ìœ„ì ¯ ì´ë²¤íŠ¸ ì—°ê²° ë° ì„¤ì •
        self.btnSend.clicked.connect(self.generate_response)
        self.lblAnswer.setReadOnly(True)

        self.chat_history = []
        self.gemini_worker = None 
        self.speech_worker = None
        
        self.lblAnswer.setText("[Mygemini] ë¬´ì—‡ì„ ë„ì™€ë“œë¦´ê¹Œìš”?")
        self.lblAnswer.append("\n[DB] ëŒ€í™” ë‚´ìš©ì´ ìë™ìœ¼ë¡œ ê¸°ë¡ë©ë‹ˆë‹¤. ì§ˆë¬¸ ì‹œ ë¨¼ì € DBì—ì„œ ê²€ìƒ‰í•©ë‹ˆë‹¤.")

        self.show()

    def generate_response(self):
        user_question = self.lineEditMyQuestion.text().strip()
        
        if not user_question:
            return
        
        # ğŸŒŸ ì‚¬ìš©ìê°€ íŠ¹ì • ëª…ë ¹ì–´ë¥¼ ì…ë ¥í•˜ë©´ ê²€ìƒ‰ ê¸°ëŠ¥ ì‹¤í–‰ (ì´ì „ ë¡œì§ ì œê±°)
        # if user_question.lower().startswith("ê²€ìƒ‰:"):
        #     search_term = user_question[3:].strip()
        #     self.search_history(search_term)
        #     self.lineEditMyQuestion.clear()
        #     return
        
        # 1. DBì—ì„œ ë¨¼ì € ê²€ìƒ‰ ì‹œë„. ê¸°ë¡ì´ ìˆìœ¼ë©´ True ë°˜í™˜ ë° UIì— í‘œì‹œ í›„ ì¢…ë£Œ.
        if self.search_history(user_question):
            self.lineEditMyQuestion.clear()
            return
            
        # 2. DBì— ê¸°ë¡ì´ ì—†ìœ¼ë©´ Gemini API í˜¸ì¶œ ì‹œì‘
        user_message = f"[ì§ˆë¬¸] {user_question}\n"
        self.lblAnswer.append(user_message)
        
        loading_message = "[Mygemini] ì‘ë‹µì„ ìƒì„±í•˜ëŠ” ì¤‘ì…ë‹ˆë‹¤..."
        self.lblAnswer.append(loading_message)
        QtWidgets.QApplication.processEvents() 
        
        self.chat_history.append({"role": "user", "parts": [{"text": user_question}]})

        # Gemini API í˜¸ì¶œ (QThread ì‚¬ìš©)
        self.gemini_worker = GeminiWorker(
            client=client,
            model_name=MODEL_NAME,
            chat_history=self.chat_history,
            user_question=user_question
        )
        
        self.gemini_worker.response_ready.connect(self.handle_response)
        self.gemini_worker.error_occurred.connect(self.handle_error)
        
        self.gemini_worker.start()
        
        self.lineEditMyQuestion.clear()
        
    def handle_response(self, user_question, gemini_response):
        """API ì‘ë‹µì„ ë°›ì•„ UIì— í‘œì‹œí•˜ê³ , DBì— ì €ì¥í•˜ë©°, ìŒì„± ì¶œë ¥í•©ë‹ˆë‹¤."""
        
        formatted_output = f"[Mygemini] {gemini_response}\n"
        
        # ë¡œë”© ë©”ì‹œì§€ë¥¼ ì œê±°í•©ë‹ˆë‹¤.
        cursor = self.lblAnswer.textCursor()
        cursor.movePosition(QtGui.QTextCursor.MoveOperation.End)
        cursor.movePosition(QtGui.QTextCursor.MoveOperation.StartOfLine, QtGui.QTextCursor.MoveMode.KeepAnchor)
        cursor.removeSelectedText()
        
        # ìµœì¢… ì‘ë‹µì„ ì¶”ê°€í•©ë‹ˆë‹¤.
        self.lblAnswer.append(formatted_output)
        
        # ğŸŒŸ DBì— ì§ˆë¬¸ê³¼ ë‹µë³€ì„ ì €ì¥í•©ë‹ˆë‹¤.
        self.save_to_mysql(user_question, gemini_response)

        # ìŒì„± ì¶œë ¥ì„ ìƒˆ ìŠ¤ë ˆë“œì—ì„œ ì‹œì‘í•©ë‹ˆë‹¤.
        self.start_speech_worker(gemini_response)

        # ìŠ¤ë ˆë“œ ì¢…ë£Œ ë° ì •ë¦¬
        self.gemini_worker.quit()
        self.gemini_worker.wait()
        
    def handle_error(self, error_type, error_message):
        """API ì˜¤ë¥˜ ë°œìƒ ì‹œ UIì— ì˜¤ë¥˜ ë©”ì‹œì§€ë¥¼ í‘œì‹œí•˜ê³  ë¡œë”© ë©”ì‹œì§€ë¥¼ ì œê±°í•©ë‹ˆë‹¤."""
        
        if self.chat_history and self.chat_history[-1].get("role") == "user":
             self.chat_history.pop()

        formatted_output = f"[Mygemini] {error_type} ë°œìƒ: {error_message}\n"
        
        # ë¡œë”© ë©”ì‹œì§€ë¥¼ ì œê±°í•˜ê³  ì˜¤ë¥˜ ë©”ì‹œì§€ë¥¼ ì¶”ê°€í•©ë‹ˆë‹¤.
        cursor = self.lblAnswer.textCursor()
        cursor.movePosition(QtGui.QTextCursor.MoveOperation.End)
        cursor.movePosition(QtGui.QTextCursor.MoveOperation.StartOfLine, QtGui.QTextCursor.MoveMode.KeepAnchor)
        cursor.removeSelectedText()
        
        self.lblAnswer.append(formatted_output)
        
        # ì˜¤ë¥˜ ë©”ì‹œì§€ ìŒì„± ì¶œë ¥ì„ ìƒˆ ìŠ¤ë ˆë“œì—ì„œ ì‹œì‘í•©ë‹ˆë‹¤.
        self.start_speech_worker(f"{error_type} ë°œìƒ: {error_message}")

        # ìŠ¤ë ˆë“œ ì¢…ë£Œ ë° ì •ë¦¬
        self.gemini_worker.quit()
        self.gemini_worker.wait()

    def start_speech_worker(self, text):
        """ë¸”ë¡œí‚¹ë  ìˆ˜ ìˆëŠ” playsoundë¥¼ ë³„ë„ì˜ SpeechWorker ìŠ¤ë ˆë“œì—ì„œ ì‹¤í–‰í•©ë‹ˆë‹¤."""
        if self.speech_worker and self.speech_worker.isRunning():
            self.speech_worker.wait() 
            
        self.speech_worker = SpeechWorker(text)
        self.speech_worker.start()

    # =================================================================
    # ğŸŒŸ ì¶”ê°€ ê¸°ëŠ¥ 1: MySQLì— ë°ì´í„° ì €ì¥ (DB ì™„ì„±)
    # =================================================================
    def save_to_mysql(self, question, answer):
        """ì§ˆë¬¸ê³¼ ë‹µë³€ì„ MySQL ë°ì´í„°ë² ì´ìŠ¤ì— ì €ì¥í•©ë‹ˆë‹¤."""
        try:
            current_time = datetime.now().strftime('%Y-%m-%d %H:%M:%S')

            # ğŸš¨ DB ì—°ê²° ì •ë³´: ì‹¤ì œ ì •ë³´ë¡œ êµì²´í•´ì•¼ í•©ë‹ˆë‹¤!
            conn = pymysql.connect(
                host='bitnmeta2.synology.me',
                user='iyrc',
                passwd='Dodan1004!',
                db='gemini_ai',
                charset='utf8',
                port=3307,
                cursorclass=pymysql.cursors.DictCursor
            )
            with conn.cursor() as cursor:
                # ğŸš¨ í…Œì´ë¸” ì´ë¦„ í™•ì¸: ì‹¤ì œ MySQL í…Œì´ë¸” ì´ë¦„ìœ¼ë¡œ ë°”ê¿”ì£¼ì„¸ìš”!
                sql = "INSERT INTO chat_history (question, answer, create_at) VALUES (%s, %s, %s)"
                
                # ì¿¼ë¦¬ ì‹¤í–‰
                cursor.execute(sql, (question, answer, current_time))
            
            conn.commit()
            print(f"âœ… MySQL ì €ì¥ ì„±ê³µ: {current_time}")

        except Exception as e:
            print(f"âŒ MySQL ì €ì¥ ì‹¤íŒ¨: {e}")
            # UIì— ì˜¤ë¥˜ ë©”ì‹œì§€ ì¶”ê°€
            self.lblAnswer.append(f"[DB ì˜¤ë¥˜] ê¸°ë¡ ì €ì¥ ì‹¤íŒ¨: {e}")
        
        finally:
            if 'conn' in locals() and conn.open:
                conn.close()

    # =================================================================
    # ğŸŒŸ ì¶”ê°€ ê¸°ëŠ¥ 2: MySQLì—ì„œ íˆìŠ¤í† ë¦¬ ê²€ìƒ‰ (ì €ì¥ëœ ë‹µ ì°¾ê¸°)
    # =================================================================
    def search_history(self, search_term):
        """MySQL ë°ì´í„°ë² ì´ìŠ¤ì—ì„œ íŠ¹ì • ë‹¨ì–´ë¥¼ í¬í•¨í•˜ëŠ” ì§ˆë¬¸ê³¼ ë‹µë³€ ê¸°ë¡ì„ ê²€ìƒ‰í•©ë‹ˆë‹¤.
           ê²€ìƒ‰ëœ ê¸°ë¡ì´ ìˆìœ¼ë©´ True, ì—†ìœ¼ë©´ Falseë¥¼ ë°˜í™˜í•©ë‹ˆë‹¤."""
        
        # ê²€ìƒ‰ ì‹œì‘ ë©”ì‹œì§€ë¥¼ UIì— ì¶”ê°€ (DB ê²€ìƒ‰ ì¤‘ì„ì„ ì•Œë¦¼)
        self.lblAnswer.append(f"\n[DB ê²€ìƒ‰] '{search_term}'ìœ¼ë¡œ ê³¼ê±° ê¸°ë¡ ê²€ìƒ‰ ì‹œì‘...")
        QtWidgets.QApplication.processEvents()
        
        search_success = False # ê²€ìƒ‰ ì„±ê³µ ì—¬ë¶€ë¥¼ ì €ì¥í•  í”Œë˜ê·¸
        
        try:
            # ğŸš¨ DB ì—°ê²° ì •ë³´: save_to_mysqlê³¼ ë™ì¼í•œ ì •ë³´ ì‚¬ìš©
            conn = pymysql.connect(
                host='bitnmeta2.synology.me',
                user='iyrc',
                passwd='Dodan1004!',
                db='gemini_ai',
                charset='utf8',
                port=3307,
                cursorclass=pymysql.cursors.DictCursor
            )

            with conn.cursor() as cursor:
                # ğŸš¨ í…Œì´ë¸” ì´ë¦„ í™•ì¸: ì‹¤ì œ MySQL í…Œì´ë¸” ì´ë¦„ìœ¼ë¡œ ë°”ê¿”ì£¼ì„¸ìš”!
                # questionê³¼ answer í•„ë“œì—ì„œ ê²€ìƒ‰ì–´ì™€ ì¼ì¹˜í•˜ëŠ” ê²ƒì„ ì°¾ìŠµë‹ˆë‹¤.
                sql = "SELECT create_at, question, answer FROM chat_history WHERE question LIKE %s OR answer LIKE %s ORDER BY create_at DESC"
                search_pattern = f"%{search_term}%" # LIKE ê²€ìƒ‰ì„ ìœ„í•œ íŒ¨í„´
                
                cursor.execute(sql, (search_pattern, search_pattern))
                results = cursor.fetchall()
            
            # ê²€ìƒ‰ ê²°ê³¼ë¥¼ UIì— í‘œì‹œ
            if results:
                search_success = True # ê²€ìƒ‰ ì„±ê³µ í”Œë˜ê·¸ ì„¤ì •
                
                self.lblAnswer.append(f"[DB ê²°ê³¼] ì´ {len(results)}ê±´ì˜ ê´€ë ¨ ê¸°ë¡ì„ ì°¾ì•˜ìŠµë‹ˆë‹¤:")
                for i, row in enumerate(results):
                    self.lblAnswer.append(f"--- [ê¸°ë¡ {i+1}] {row['create_at']} ---")
                    self.lblAnswer.append(f"  Q: {row['question'][:50]}...") # ì§ˆë¬¸ì€ 50ìë§Œ ë¯¸ë¦¬ë³´ê¸°
                    self.lblAnswer.append(f"  A: {row['answer'][:50]}...") # ë‹µë³€ë„ 50ìë§Œ ë¯¸ë¦¬ë³´ê¸°
                self.lblAnswer.append("--------------------------------------------------")
            else:
                self.lblAnswer.append("[DB ê²°ê³¼] í•´ë‹¹ ê²€ìƒ‰ì–´ì™€ ì¼ì¹˜í•˜ëŠ” ê³¼ê±° ê¸°ë¡ì´ ì—†ìŠµë‹ˆë‹¤. Geminiì— ì§ˆë¬¸í•©ë‹ˆë‹¤.")
            
            print(f"âœ… MySQL ê²€ìƒ‰ ì„±ê³µ: {len(results)}ê±´")
            
            return search_success # ìµœì¢… ê²°ê³¼ ë°˜í™˜

        except Exception as e:
            print(f"âŒ MySQL ê²€ìƒ‰ ì‹¤íŒ¨: {e}")
            self.lblAnswer.append(f"[DB ì˜¤ë¥˜] ê¸°ë¡ ê²€ìƒ‰ ì‹¤íŒ¨: {e}")
            # ì˜¤ë¥˜ ë°œìƒ ì‹œì—ë„ Gemini í˜¸ì¶œì„ ìœ„í•´ False ë°˜í™˜ (DB ë¬¸ì œê°€ ìˆë‹¤ë©´ AIê°€ ì‘ë‹µí•´ì•¼ í•¨)
            return False 
        
        finally:
            if 'conn' in locals() and conn.open:
                conn.close()


# =================================================================
# 4. í”„ë¡œê·¸ë¨ ì‹¤í–‰
# =================================================================

if __name__ == '__main__':
    # í”„ë¡œê·¸ë¨ ì‹œì‘ ì‹œ ë‚¨ì•„ìˆëŠ” temp_speech_*.mp3 íŒŒì¼ì„ ì‚­ì œí•©ë‹ˆë‹¤.
    for f in os.listdir('.'):
        if f.startswith('temp_speech_') and f.endswith('.mp3'):
            try:
                os.remove(f)
            except OSError as e:
                print(f"ì„ì‹œ íŒŒì¼ ì‚­ì œ ì˜¤ë¥˜ {f}: {e}")
                
    app = QtWidgets.QApplication(sys.argv)
    window = GeminiApp()
    sys.exit(app.exec())